File: solutions/problem_2/solution.rb
def sum_even_fibonacci_numbers
  sum = 0
  first_item = 1
  second_item = 1
  third_item = first_item+second_item
  while third_item < 4000000
    sum += third_item
    first_item = second_item + third_item
    second_item = first_item + third_item
    third_item = first_item + second_item
  end
  sum
end

require './ruby/benchmark'
ProjectEuller::RubyBenchmark.run { sum_even_fibonacci_numbers }

File: solutions/problem_2/solution.ex
defmodule Benchmark do
  def measure(function) do
    function
    |> :timer.tc
    |> elem(0)
    |> Kernel./(1_000_000)
    |> Float.to_string([decimals: 10, compact: true])
  end
end

defmodule ProblemTwo do

  def sum_fibonacci(_, second, third, sum) when third < 4000000 do
    new_first = second + third
    new_second = new_first + third
    new_third = new_first + new_second

    sum_fibonacci(new_first, new_second, new_third, sum + third)
  end

  def sum_fibonacci(_, _, _, sum), do: sum

  def sum_even_fibonacci_numbers do
    IO.puts "=> Result: #{sum_fibonacci(1, 1, 2, 0)}"
    IO.puts "=> Time: #{Benchmark.measure(fn -> sum_fibonacci(1, 1, 2, 0) end)}"
  end
end

ProblemTwo.sum_even_fibonacci_numbers